l3kernel/l3expan.dtx

```diff
@@ -125,87 +130,92 @@
+%   |\foo:Nn|. Similarly
+%   \begin{quote}
+%     |\cs_generate_variant:Nn \foo:Nn { NV , cV }|
+%   \end{quote}
+%   generates the variants |\foo:NV| and |\foo:cV| in the same
+%   way. Notice that
+%   \begin{quote}
+%     |\cs_generate_variant:Nn \foo:cn { cV }|
+%   \end{quote}
+%   is an error to generate |\foo:cV| because |cn| is not a
+%   \meta{base arg spec}. On the contrary
+%   \begin{quote}
+%     |\cs_generate_variant:Nn \foo:nc { Vc }|
+%   \end{quote}
+%   is allowed because |c| is not part of the \meta{base arg spec} head.
+%
+%   The \cs{cs_generate_variant:Nn} function should only be applied to
+%   already defined base functions\footnotemark{}.
+%   If the base function is protected or if the \meta{variant arg spec} contains
+%   |x|, then the variant is
+%   protected. There is no need to re-apply \cs{cs_generate_variant:Nn} after
+%   changing the definition of the base function: the variant will always
+%   use the current definition of the base.
+%   As \cs{cs_generate_variant:Nn} does nothing when the variant
+%   is already available, generating the same variants repeatedly is 
+%   safe.
 %
 %   For backward compatibility it is currently possible to make |n|,
-%   |o|, |V|, |v|, |f|, |e|, or |x|-type variants of an |N|-type argument or
-%   |N| or |c|-type variants of an |n|-type argument.  Both are
+%   |o|, |V|, |v|, |f|, |e|, or |x| type variants of an |N| type argument or
+%   |N| or |c| type variants of an |n| type argument.  Both are
 %   deprecated.  The first because passing more than one token to an
-%   |N|-type argument will typically break the parent function's code.
+%   |N| type argument will typically break the parent function's code.
 %   The second because programmers who use that most often want to
 %   access the value of a variable given its name, hence should use a
-%   |V|-type or |v|-type variant instead of |c|-type.  In those cases,
+%   |V| type or |v| type variant instead of |c| type.  In those cases,
 %   using the lower-level \cs{exp_args:No} or \cs{exp_args:Nc}
 %   functions explicitly is preferred to defining confusing variants.
 % \end{function}
+% \footnotetext{This is only
+%   enforced if debugging support \texttt{check-declarations} is enabled.}
 %
 % \begin{function}[added = 2018-04-04, updated = 2019-02-08]
 %   {\exp_args_generate:n}
 %   \begin{syntax}
 %     \cs{exp_args_generate:n} \Arg{variant argument specifiers}
 %   \end{syntax}
-%   Defines \cs[no-index]{exp_args:N\meta{variant}} functions for each
-%   \meta{variant} given in the comma list \Arg{variant argument
-%   specifiers}.  Each \meta{variant} should consist of the letters |N|,
+%   Defines \cs[no-index]{exp_args:N\meta{variant arg spec}} functions for each
+%   \meta{variant arg spec} given in the comma list \meta{variant argument
+%   specifiers}.  Each \meta{variant arg spec} should consist of the letters |N|,
 %   |c|, |n|, |V|, |v|, |o|, |f|, |e|, |x|, |p| and the resulting function is
```

Comment:
`\exp_args:N<variant arg spec>` functions can have `p`-type arguments?

Reply:
I missed that. This was very unclear in the sense that originally, the doc reads

> Only n and N arguments can be changed to other types. The only allowed changes are
> • c variant of an N parent;
> • o,V,v,f,e,orx variant of an n parent;
> • N, n, T, F, or p argument unchanged.

The last line implicitly allows |p| arguments, but that contradicts the notion of base function: |p| should be removed.

Reply:
I meant, since functions like `\exp_args:Npe` will never exist, `p` should be removed from the `<variant arg spec>` list here, on line 211.

Reply:
Yes, "p" should definitely be removed from that list.

Reply:
On second thoughts, I can see why it is there.  Maybe there should be a second list (p, T, F) of those that can only be used as placeholders and that never indicate replacements (i.e., they must always indicate "no change here")?

Reply:
I was wrong. `\exp_args:Npe` is definable and it actually works. Note that here `p` denotes the `<parameter text>` in `\cs_new:Npn` and friends.